### PR TITLE
feat: anti join

### LINKED ISSUE #374: [Enh]: Support for anti-join
### We would like to learn about your use case. For example, if this feature is needed  to adopt Narwhals in an open source project, could you please enter the link to it below?

_No response_

### Please describe the purpose of the new feature or describe the problem to solve.

Hands down one of my favorite feature 😂
The fact that polars has native support is really convenient

### Suggest a solution if possible.

It should look something like the following:

```py
(df
.merge(other, how='outer', indicator = True)
.loc[lambda t: t["_merge"]=="left_only"]
.drop("_merge", axis=1)
)
```

We can decide to name the indicator column with another non-default value

### If you have tried alternatives, please describe them below.

_No response_

### Additional information that may help us understand your needs.

_No response_

### PR DESCRIPTION
**Pull request by FBruzzesi**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #374 

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added 
- [x] Documented the changes

### CODE REVIEW

File: narwhals/_pandas_like/dataframe.py [hunk 1]
```diff
@@ -295,27 +296,9 @@ def join(
...
-                        if counter > 100:  # pragma: no cover
-                            msg = (
-                                "Internal Error: Narwhals was not able to generate a column name to perform cross "
-                                "join operation"
-                            )
-                            raise AssertionError(msg)
-
-                key_token = generate_unique_token(8, self.columns)
+                key_token = generate_unique_token(
+                    n_bytes=8, columns=[*self.columns, *other.columns]
```
Comment (FBruzzesi) [inline narwhals/_pandas_like/dataframe.py:300]: Nobody noticed, but I think adding `other.columns` is a safer way

File: narwhals/_pandas_like/dataframe.py [hunk 2]
```diff
@@ -335,6 +318,23 @@ def generate_unique_token(
...
                 )
 
+        if how == "anti":
+            indicator_token = generate_unique_token(
+                n_bytes=8, columns=[*self.columns, *other.columns]
+            )  # pragma: no cover
+
+            return self._from_dataframe(
+                self._dataframe.merge(
+                    other._dataframe.loc[:, left_on],
```
Comment (FBruzzesi) [inline narwhals/_pandas_like/dataframe.py:328]: selecting only the join columns in here so that no additional columns get created

File: tests/frame/test_common.py [hunk 3]
```diff
@@ -325,36 +325,29 @@ def test_join(df_raw: Any) -> None:
...
-        ),
-    ],
-)
-def test_cross_join(df_raw: Any, expected: dict[str, list[Any]]) -> None:
+@pytest.mark.parametrize("df_raw", [df_polars, df_lazy, df_pandas, df_mpd])
+def test_cross_join(df_raw: Any) -> None:
     df = nw.from_native(df_raw).select("a")
     result = df.join(df, how="cross")  # type: ignore[arg-type]
+
+    expected = {"a": [1, 1, 1, 3, 3, 3, 2, 2, 2], "a_right": [1, 3, 2, 1, 3, 2, 1, 3, 2]}
```
Comment (FBruzzesi) [inline tests/frame/test_common.py:333]: I think it was parametrized because I wanted to include tests with None values?!

**Comment by FBruzzesi** [comment]
It seems I can replicate the error when running with pytest, but can't on ipython  🤔 I am getting somewhat confused

```py
import modin.pandas as mpd
import narwhals as nw

df = mpd.DataFrame({"a": [1, 3, 2], "b": [4, 4, 6], "z": [7.0, 8, 9]})
df_ = nw.from_native(df)
other = df_.filter(nw.col("b") < 6)

result = df_.join(other, how="anti", left_on=["a", "b"], right_on=["a", "b"])
nw.to_native(result)
```

image

**Comment by MarcoGorelli** [comment]
I think in ci we have `modin[dask]`, looks like you have `modin[ray]`?

File: narwhals/_pandas_like/dataframe.py (hunk 1 above)
Comment (MarcoGorelli) [reply narwhals/_pandas_like/dataframe.py:300]: ah, nice! indeed, I hadn't noticed 🙈 thanks for picking up on this

**Comment by FBruzzesi** [comment]
I tried both backends, ending up with the same result:

- failing tests
- non failing ipython

File: narwhals/_pandas_like/dataframe.py [hunk 4]
```diff
@@ -335,6 +318,25 @@ def generate_unique_token(
...
+                        columns=dict(zip(right_on, left_on))  # type: ignore[arg-type]
+                    ),
+                    how="outer",
+                    indicator=indicator_token,
+                    left_on=left_on,
+                    right_on=left_on,
+                )
+                .loc[lambda t: t[indicator_token] == "left_only"]
+                .drop(columns=[indicator_token])
+            )
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/dataframe.py:338]: I'm not 100% sure this is the same - could you include a test in which you anti-join on a single column which contains duplicates please?

Comment (FBruzzesi) [reply narwhals/_pandas_like/dataframe.py:338]: Just to give you a snippet of intermediate values:

```py
other = df_raw.loc[lambda t: t["b"]>5, "b"]
df_raw.merge(other, how="outer", indicator=True, left_on=["b"], right_on=["b"] )
```
```terminal
   a  b    z     _merge
0  1  4  7.0  left_only
1  3  4  8.0  left_only
2  2  6  9.0       both
```
```py
other = df_raw.loc[lambda t: t["b"]<5, "b"]
df_raw.merge(other, how="outer", indicator=True, left_on=["b"], right_on=["b"])
```

```terminal
   a  b    z     _merge
0  1  4  7.0       both
1  1  4  7.0       both
2  3  4  8.0       both
3  3  4  8.0       both
4  2  6  9.0  left_only
```

The latter could blow up memory?!
That said, I personally use anti join only if I need to filter on multiple columns i.e. tuple of values, otherwise `~ Expr.is_in(...)` is all you need.

Comment (MarcoGorelli) [reply narwhals/_pandas_like/dataframe.py:338]: ok thanks - maybe this is correct then (I think it's for semi-join that duplicates may introduce issues)

but yes, I think there is a risk of memory blow-up, though arguably that's always a risk with pandas...maybe we just need a note cautioning users about using this?

Comment (FBruzzesi) [reply narwhals/_pandas_like/dataframe.py:338]: Should we drop duplicates ourselves in `other`?

```py
(other._dataframe
 .loc[:, right_on]
.rename(columns=dict(zip(right_on, left_on)))
.drop_duplicates(ignore_index=True)
)
```

File: tests/conftest.py [hunk 5]
```diff
@@ -58,7 +57,7 @@ def polars_constructor(obj: Any) -> IntoDataFrame:
 else:  # pragma: no cover
     params = [pandas_constructor]
 params.append(polars_constructor)
-if os.environ.get("CI") and get_modin() is not None:  # pragma: no cover
+if get_modin() is not None:  # pragma: no cover
```
Comment (FBruzzesi) [inline tests/conftest.py:60]: As commented in #379, mind keeping it?!

File: narwhals/_pandas_like/dataframe.py (hunk 4 above)
Comment (MarcoGorelli) [reply narwhals/_pandas_like/dataframe.py:338]: yeah that's probably be better 👍

File: narwhals/_pandas_like/dataframe.py [hunk 6]
```diff
@@ -335,6 +318,31 @@ def generate_unique_token(
...
+            other_ = (
+                other._dataframe.loc[:, right_on]
+                .rename(  # rename to avoid creating extra columns in join
+                    columns=dict(zip(right_on, left_on))  # type: ignore[arg-type]
+                )
+                .drop_duplicates(ignore_index=True)
+            )
+            return self._from_dataframe(
+                self._dataframe.merge(
+                    other_,
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/dataframe.py:335]: why nnot just reassign to `other`?
```suggestion
            other = (
                other._dataframe.loc[:, right_on]
                .rename(  # rename to avoid creating extra columns in join
                    columns=dict(zip(right_on, left_on))  # type: ignore[arg-type]
                )
                .drop_duplicates(ignore_index=True)
            )
            return self._from_dataframe(
                self._dataframe.merge(
                    other,
```

File: narwhals/_pandas_like/dataframe.py [hunk 7]
```diff
@@ -335,6 +318,31 @@ def generate_unique_token(
                     ),
                 )
 
+        if how == "anti":
+            indicator_token = generate_unique_token(
+                n_bytes=8, columns=[*self.columns, *other.columns]
+            )  # pragma: no cover
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/dataframe.py:324]: love this

File: tests/conftest.py (hunk 5 above)
Comment (MarcoGorelli) [reply tests/conftest.py:60]: sure!

**Review by MarcoGorelli — APPROVED** [review]
thanks @FBruzzesi ! left a minor comment but should be good, feel free to merge

File: narwhals/_pandas_like/dataframe.py [hunk 8]
```diff
@@ -295,27 +296,9 @@ def join(
...
-                            msg = (
-                                "Internal Error: Narwhals was not able to generate a column name to perform cross "
-                                "join operation"
-                            )
-                            raise AssertionError(msg)
-
-                key_token = generate_unique_token(8, self.columns)
+                key_token = generate_unique_token(
+                    n_bytes=8, columns=[*self.columns, *other.columns]
+                )  # pragma: no cover
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/dataframe.py:301]: do we need this? doesn't the `._implemention = 'modin'` in one of the tests which run by default hit this?

**Comment by MarcoGorelli** [comment]
thanks @FBruzzesi !
